internal/pkg/skuba/addons/gangway.go

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2019 SUSE LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package addons
+
+import (
+	"github.com/SUSE/skuba/internal/pkg/skuba/gangway"
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubernetes"
+	"github.com/SUSE/skuba/internal/pkg/skuba/skuba"
+	skubaconstants "github.com/SUSE/skuba/pkg/skuba"
+	"github.com/pkg/errors"
+	"k8s.io/kubernetes/cmd/kubeadm/app/images"
+)
+
+func init() {
+	registerAddon(kubernetes.Gangway, renderGangwayTemplate, GangwayCallbacks{})
+}
+
+func (addonRenderContext addonRenderContext) GangwayImage() string {
+	return images.GetGenericImage(skubaconstants.ImageRepository, "gangway",
+		kubernetes.AddonVersionForClusterVersion(kubernetes.Gangway, addonRenderContext.Config.ClusterVersion).Version)
+}
+
+func renderGangwayTemplate(addonConfiguration AddonConfiguration) string {
+	return gangwayManifest
+}
+
+type GangwayCallbacks struct{}
+
+func (GangwayCallbacks) BeforeApply(addonConfiguration AddonConfiguration, skubaConfiguration *skuba.SkubaConfiguration) error {
```

Comment:
The implementation of this function looks really similar to the dex one (with the difference on the secrets part). I'm wondering if we could share more code. This can be done on a later PR whenever we have more addons, though.

Reply:
Could be, however I wanted as a feature on this PR to leave every addon logic separate from each other. We can improve in future PR's though.